abstractions/common/modules.py

```diff
@@ -168,45 +168,120 @@ def forward(self, state):
+            final_size = conv2d_size_out(final_size, (3, 3), 1)
+            output_size = final_size[0] * final_size[1] * 64
+
+            self.body1 = torch.nn.Sequential(
+                torch.nn.Conv2d(num_frames, 32, kernel_size=(8, 8), stride=4),
+                torch.nn.ReLU(),
+                torch.nn.Conv2d(32, 64, kernel_size=(4, 4), stride=2),
+                torch.nn.ReLU(),
+                torch.nn.Conv2d(64, 64, kernel_size=(3, 3), stride=1),
+                torch.nn.ReLU(),
+                Reshape(-1, output_size))
+            self.q1 = torch.nn.Sequential(
+                torch.nn.Linear(output_size + num_actions, hidden_dim),
+                torch.nn.ReLU(),
+                torch.nn.Linear(hidden_dim, 1))
+
+            self.body2 = torch.nn.Sequential(
+                torch.nn.Conv2d(num_frames, 32, kernel_size=(8, 8), stride=4),
+                torch.nn.ReLU(),
+                torch.nn.Conv2d(32, 64, kernel_size=(4, 4), stride=2),
+                torch.nn.ReLU(),
+                torch.nn.Conv2d(64, 64, kernel_size=(3, 3), stride=1),
+                torch.nn.ReLU(),
+                Reshape(-1, output_size))
+            self.q2 = torch.nn.Sequential(
+                torch.nn.Linear(output_size + num_actions, hidden_dim),
+                torch.nn.ReLU(),
+                torch.nn.Linear(hidden_dim, 1))
 
         self.apply(weights_init_)
 
-    def forward(self, state, action):
-        xu = torch.cat([state, action], 1)
-
-        x1 = torch.nn.functional.relu(self.linear1(xu))
-        x1 = torch.nn.functional.relu(self.linear2(x1))
-        x1 = self.linear3(x1)
-
-        x2 = torch.nn.functional.relu(self.linear4(xu))
-        x2 = torch.nn.functional.relu(self.linear5(x2))
-        x2 = self.linear6(x2)
+        trainable_parameters = sum(p.numel() for p in self.parameters() if p.requires_grad)
+        print(repr(self))
+        print("Number of trainable parameters: {}".format(trainable_parameters))
 
+    def forward(self, state, action):
+        if self.model_type == 'cnn':
+            x1 = self.q1(torch.cat([self.body1(state), action], 1))
+            x2 = self.q2(torch.cat([self.body2(state), action], 1))
+        elif self.model_type == 'mlp':
+            xu = torch.cat([state, action], 1)
+            x1 = self.q1(xu)
+            x2 = self.q2(xu)
         return x1, x2
 
 
 class GaussianPolicy(torch.nn.Module):
-    def __init__(self, num_inputs, num_actions, hidden_dim, action_space=None):
+    def __init__(self,
+                 input_shape,
+                 num_actions,
+                 hidden_dim,
+                 model_type,
+                 num_frames=None,
+                 action_space=None):
         super(GaussianPolicy, self).__init__()
 
-        self.linear1 = torch.nn.Linear(num_inputs, hidden_dim)
-        self.linear2 = torch.nn.Linear(hidden_dim, hidden_dim)
+        if len(input_shape) > 2:
+            input_shape = input_shape[1:]
```

Comment:
Is this what makes it so you don't need `env.observation_space.shape[0]` anymore?

Reply:
No this makes the framestacking work with the batching, so that the cnn final_output_size isn't confused.